keps/sig-apps/3939-include-terminating-pods-as-active/README.md

```diff
@@ -0,0 +1,781 @@
+do utilize resources because resources are still allocated to them and there is potential for a user to be charged for utilizing those resources.
+
+In scarce compute environments, these resources can be difficult to obtain so pods can take a long time to find resources and they may only be able to find nodes once the existing pods have been terminated.
+
+### Goals
+
+- Job Controller should only create new pods once the existing ones are marked as Failed/Succeeded
+- Deployment controller should allow for flexibility in waiting for pods to be fully terminated before
+  creating new ones
+
+### Non-Goals
+
+- DaemonSets and StatefulSets are not included in this proposal
+  - They were designed to enforce uniqueness from the start so we will not include them in this design.
+
+## Proposal
+
+Both Jobs and the ReplicaSet controller get a list of active pods.  Active pods usually mean pods that have not been registered for deletion.  In this KEP, we want to include terminating pods as active pods.
+
+We will propose two new API fields in Jobs and Deployments/ReplicaSets in this KEP.  
+
+### User Stories (Optional)
+
+#### Story 1
+
+As a machine learning user, ML frameworks allow scheduling of multiple pods.  
+The Job controller does not typically wait for terminating pods to be marked as failed.  Tensorflow and other ML frameworks may have a requirement that they only want Pods to be started once the other pods are fully terminated.  The following yaml can fit these needs:
+
+This case was added due to a bug discovered with running IndexedJobs with Tensorflow.  See [Jobs create replacement Pods as soon as a Pod is marked for deletion](https://github.com/kubernetes/kubernetes/issues/115844) for more details.
+
+#### Story 2
+
+As a cloud user, users would want to guarantee that the number of pods that are running is exactly the amount that they specify.  Terminating pods do not relinguish resources so scarce compute resource are still scheduled to those pods.
+See [Kueue: Account for terminating pods when doing preemption](https://github.com/kubernetes-sigs/kueue/issues/510) for an example of this.
+
+#### Story 3
+
+As a cloud user, users would want to guarantee that the number of pods that are running includes terminating pods.  In scare compute environments, users may only have a limited amount of nodes and they do not want to try and schedule pods to a new resource.
+Counting terminating pods as active allows for the scheduling of pods to wait until pods are terminated.
+
+See [Option for acknowledging terminating Pods in Deployment rolling update](https://github.com/kubernetes/kubernetes/issues/107920)
+for more examples.
+
+### Notes/Constraints/Caveats (Optional)
+
+The Deployment API is open for discussion.  We put the field in Deployment/ReplicaSet because it is related to RolloutStrategy.
+It is not clear if `recreate` and/or `rollingupdate` need this API for both rollout options.
```

Comment:
By default, I would expect both strategies to have the behavior unless a very strong case could be made that one of them shouldn’t have it.  In general, i would not expect this behavior to be tied to the update strategy, but fundamental to the object, since replacement is not considered part of update strategy and this behavior would impact replacement.

Reply:
Agreed on scale.  Spec level seems reasonable with a consistent name across all types that have a concept of minReadySeconds

Reply:
I'm afraid I don't know very much about the scale.  Is this still covered by DeploymentSpec?

Reply:
Scaling is achieved by calling the scale API or changing the replicas of the deployment. Either way the rest is handled internally by the controller.

Reply:
As mentioned we decided to move the deployment discussion to a new KEP so we can roll these KEPs out in different candences.  

https://github.com/kubernetes/enhancements/pull/3974